linkerd/proxy/http/src/balance.rs

```diff
@@ -1,59 +1,88 @@
+    pub default_rtt: Duration,
+    pub decay: Duration,
+}
 
 /// Configures a stack to resolve `T` typed targets to balance requests over
 /// `M`-typed endpoint stacks.
 #[derive(Debug)]
-pub struct Layer<A, B> {
-    decay: Duration,
-    default_rtt: Duration,
-    _marker: PhantomData<fn(A) -> B>,
+pub struct NewBalance<ReqB, RspB, R, N> {
+    resolve: R,
+    inner: N,
+    _marker: PhantomData<fn(ReqB) -> RspB>,
 }
 
-// === impl Layer ===
+pub type Balance<B, S> =
+    p2c::Balance<PeakEwmaDiscover<discover::Buffer<S>, PendingUntilFirstData>, http::Request<B>>;
 
-pub fn layer<A, B>(default_rtt: Duration, decay: Duration) -> Layer<A, B> {
-    Layer {
-        decay,
-        default_rtt,
-        _marker: PhantomData,
-    }
-}
+// === impl NewBalance ===
 
-impl<A, B> Clone for Layer<A, B> {
-    fn clone(&self) -> Self {
+impl<ReqB, RspB, R, N> NewBalance<ReqB, RspB, R, N> {
+    // FIXME(ver)
+    const CAPACITY: usize = 1_000;
+
+    pub fn new(inner: N, resolve: R) -> Self {
         Self {
-            decay: self.decay,
-            default_rtt: self.default_rtt,
+            resolve,
+            inner,
             _marker: PhantomData,
         }
     }
+
+    pub fn layer(resolve: R) -> impl layer::Layer<N, Service = Self> + Clone
+    where
+        R: Clone,
+    {
+        layer::mk(move |inner| Self::new(inner, resolve.clone()))
+    }
 }
 
-impl<D, S, A, B> tower::layer::Layer<D> for Layer<A, B>
+impl<T, ReqB, RspB, R, M, N, S> NewService<T> for NewBalance<ReqB, RspB, R, M>
 where
-    A: HttpBody,
-    B: HttpBody,
-    D: Discover<Service = S>,
-    D::Key: Hash,
-    S: tower::Service<http::Request<A>, Response = http::Response<B>>,
+    T: Param<EwmaConfig>,
+    ReqB: HttpBody,
+    RspB: HttpBody,
+    R: Resolve<T>,
+    M: NewService<T, Service = N> + Clone,
+    N: NewService<(SocketAddr, R::Endpoint), Service = S>,
+    S: Service<http::Request<ReqB>, Response = http::Response<RspB>>,
     S::Error: Into<Error>,
-    Balance<PeakEwmaDiscover<D, PendingUntilFirstData>, http::Request<A>>:
-        tower::Service<http::Request<A>>,
+    NewSpawnDiscover<R, M>: NewService<T, Service = discover::Buffer<S>>,
+    discover::Buffer<S>: futures::Stream<Item = discover::Result<S>>,
+    Balance<ReqB, S>: tower::Service<http::Request<ReqB>>,
 {
-    type Service = Balance<PeakEwmaDiscover<D, PendingUntilFirstData>, http::Request<A>>;
+    type Service = Balance<ReqB, S>;
 
-    fn layer(&self, discover: D) -> Self::Service {
-        let instrument = PendingUntilFirstData::default();
-        let loaded = PeakEwmaDiscover::new(discover, self.default_rtt, self.decay, instrument);
-        Balance::from_rng(loaded, &mut thread_rng()).expect("RNG must be valid")
+    fn new_service(&self, target: T) -> Self::Service {
+        let EwmaConfig { default_rtt, decay } = target.param();
+        let disco = NewSpawnDiscover::new(Self::CAPACITY, self.resolve.clone(), self.inner.clone())
+            .new_service(target);
```

Comment:
style nit, take it or leave it: since the `NewSpawnDiscover` is only used once in the `NewService` impl for `the NewBalance` types...does this really need to be a whole `NewService` impl? couldn't we just have a function taking those arguments and returning a `discover::Buffer`?

or, is there an intent to compose the `NewSpawnDiscover` `NewService` with other `NewService`s later?

Reply:
really, the goal is to have our modules expose stacks so they don't dictate how they're called--it makes them composable. agree that the current usage is a bit strained.